src/java.base/share/classes/com/sun/crypto/provider/GaloisCounterMode.java

```diff
@@ -657,6 +697,72 @@ int getBufferedLength() {
             return (ibuffer == null ? 0 : ibuffer.size());
         }
 
+        /**
+         *  ByteBuffer wrapper for intrinsic implGCMCrypt
+         */
+        int implGCMCrypt(GCMOperation op, ByteBuffer src, ByteBuffer dst) {
+            int srcLen = src.remaining() - (src.remaining() % blockSize);
+            int rlen = srcLen;
+
+
+            if (srcLen < blockSize) {
+                return 0;
+            }
+
+            // 'in' and 'out' are always set together, just need to check 'in'
+            if (in == null || (in.length != PARALLEL_LEN
+                && in.length < srcLen)) {
+                in = new byte[Math.min(PARALLEL_LEN, srcLen)];
+                out = new byte[Math.min(PARALLEL_LEN, srcLen)];
+            }
+
+            byte[] ct;
+            if (encryption) {
+                ct = out;
+            } else {
+                ct = in;
+            }
+
+            if (srcLen > PARALLEL_LEN) {
+
+                if (src.hasArray() && dst.hasArray()) {
+                    int l = rlen;
+
+                    if (encryption) {
+                        rlen -= GaloisCounterMode.implGCMCrypt(src.array(),
+                            src.arrayOffset() + src.position(), src.remaining(),
+                            dst.array(), dst.arrayOffset() + dst.position(),
+                            dst.array(), dst.arrayOffset() + dst.position(),
+                            gctr, ghash);
+                    } else {
+                        rlen -= GaloisCounterMode.implGCMCrypt(src.array(),
+                            src.arrayOffset() + src.position(), src.remaining(),
+                            src.array(), src.arrayOffset() + src.position(),
+                            dst.array(), dst.arrayOffset() + dst.position(),
+                            gctr, ghash);
+                    }
```

Comment:
Can we use another ByteBuffer variable to avoid almost-duplicate calls?
```
ByteBuffer ct = (encryption? dst : src);
rlen -= GaloisCounterMode.implGCMCrypt(src.array(),
                            src.arrayOffset() + src.position(), src.remaining(),
                            ct.array(), ct.arrayOffset() + ct.position(),
                            dst.array(), dst.arrayOffset() + dst.position(),
                            gctr, ghash);
```

Reply:
That maybe a better choice